### PR TITLE
Admins can CRUD a list of Authors

### PR DESCRIPTION
**Pull request by stephaniewillhide**
#30

### CODE REVIEW

File: app/controllers/authors_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+class AuthorsController < ApplicationController
+  before_action :find_author, only: [:edit, :update, :destroy, :toggle_featured]
```
Comment (jordanmaguire) [inline app/controllers/authors_controller.rb:2]: One of these actions does not belong

File: app/controllers/authors_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  private def find_author
+    @author = Author.find(params[:id])
+  end
+
+  private def object_params
+    params.require(:author).permit(:name, :biography, :publisher_name, :publisher_email)
+  end
+end
```
Comment (jordanmaguire) [inline app/controllers/authors_controller.rb:53]: This is all looking great 👍🏻

File: app/models/author.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  validates :publisher_name, presence: true
+
+  validates :publisher_email, presence: true
+
+  scope :ordered, -> { order('LOWER(authors.name)') }
+
+  scope :search, -> (search_term) do
+    search_term_with_wildcards = "%#{search_term}%"
+    where("name LIKE ? OR publisher_name LIKE ?",
+    search_term_with_wildcards, search_term_with_wildcards)
```
Comment (jordanmaguire) [inline app/models/author.rb:19]: Indenting is a little off here - might be possible to have this one line

File: app/views/books/_form.html.erb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 <%= simple_form_for(@book) do |form| %>
   <%= form.input(:name) %>
+  <%= form.input(:author_id, collection: Author.ordered) %>
```
Comment (jordanmaguire) [inline app/views/books/_form.html.erb:3]: I don't think this works - should be a collection of authors, like genres below

File: db/migrate/20210226014754_create_authors.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      t.string :publisher_name
+      t.string :publisher_email
+      t.timestamps
+    end
+
+    add_index :authors, :name, unique: true
+
+    change_table :books do |t|
+      t.belongs_to :author, index: true, foreign_key: true
+    end
```
Comment (jordanmaguire) [inline db/migrate/20210226014754_create_authors.rb:15]: Do you still need this?

File: db/seeds.rb [hunk 6]
```diff
@@ -17,20 +32,14 @@
...
   matching_genres = genres.select { |genre| genre_names.include?(genre.name) }
   book.update(genres: matching_genres)
-  unless book.cover.attached?
-    book.cover.attach(
-      content_type: "image/jpg",
-      filename: cover.split("/").last,
-      io: File.open(Rails.root.join(cover)),
-    )
-  end
 end
```
Comment (jordanmaguire) [inline db/seeds.rb:45]: Out of curiosity, why did you remove the covers?

File: spec/models/author_spec.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    Author = Author.new
+
+    Author.valid?
+    expect(Author.errors[:name]).to eq(["can't be blank"])
+
+    Author.name = "Test Author"
+    Author.valid?
+    expect(Author.errors[:name]).to eq([])
+  end
+end
```
Comment (jordanmaguire) [inline spec/models/author_spec.rb:75]: Great 👍🏻

**Review by jordanmaguire — COMMENTED** [review]
There's a few minor things to fix up

File: db/migrate/20210226014754_create_authors.rb (hunk 5 above)
Comment (stephaniewillhide) [reply db/migrate/20210226014754_create_authors.rb]: When I removed it, the app still works. I'm assuming this is only for the has one/has many associations?

File: db/seeds.rb (hunk 6 above)
Comment (stephaniewillhide) [reply db/seeds.rb:52]: A rebase error! Returned.

**Review by jordanmaguire — APPROVED** [review]
Ship it!
